### ISSUE TITLE
Optional returns the string | undefined type even with the default value set

### ISSUE THREAD

**Issue by NotWorkingCode** [fabian-hiller/valibot#286]
I use the following schema:

```ts
export const ConfigSchema = v.object({
    logger: v.object({
        level: v.optional(v.string(), "test"),
        path: v.optional(v.nullable(v.string()), null)
    })
})
```

When I try to pass the value after validation, I get an error like:
`TS2345: Argument of type 'string | undefined' is not assignable to parameter of type 'string'.   Type 'undefined' is not assignable to type 'string'.`

Type: `level?: string | undefined`

This is a strange behavior, because when the default value is set, either the value from json or the default value will be returned to me.

**Comment by NotWorkingCode** [comment]
Thank you for your support.

`struct` is indeed set to `true`

I tried using both Input and Output - the result is the same.

If you use `Input`
![image](https://github.com/fabian-hiller/valibot/assets/52665159/dfc88237-1878-4ed6-959e-a63e37722600)

If you use `Output`
![image](https://github.com/fabian-hiller/valibot/assets/52665159/99a52faf-f937-40a8-a7aa-f4b5ed456a1a)

I am using the latest version: 0.23.0

**Comment by NotWorkingCode** [comment]
Now I also noticed that path is always equal to a string, although it can be both a string and null

**Comment by fabian-hiller** [comment]
Is `strict` set to `true` in your `tsconfig.json`?

**Comment by NotWorkingCode** [comment]
> Is `strict` set to `true` in your `tsconfig.json`?

Yes, I wrote it in my second comment.
For clarity, here is the entire tsconfig
![image](https://github.com/fabian-hiller/valibot/assets/52665159/b74fb069-7a9f-4fcb-aa79-6af8e4245dc4)

**Comment by fabian-hiller** [comment]
I don’t know what the problem is. In my playground everything works as expected.

![screenshot-code-1](https://github.com/fabian-hiller/valibot/assets/35291865/80aceeea-6399-4ad7-9203-478eb2629ec5)

![screenshot-code-2](https://github.com/fabian-hiller/valibot/assets/35291865/a710ee48-cae7-4539-bee4-fc043d680fe9)

**Comment by fabian-hiller** [comment]
Just noticed that the question marks are missing in `ConfigInput`. Will fix that in the next release.

**Comment by NotWorkingCode** [comment]
Thank you very much for your work and such a quick response.

**Comment by fabian-hiller** [comment]
v0.24.0 is available. Can you check if the problem still exists on your end?

**Comment by NotWorkingCode** [comment]
I get the same results on version 0.24.0

Input
![image](https://github.com/fabian-hiller/valibot/assets/52665159/9317f984-b77c-4021-a22d-56765878ded7)

Output
![image](https://github.com/fabian-hiller/valibot/assets/52665159/d7d7ff17-864c-45f3-a442-2fb414428912)

**Comment by fabian-hiller** [comment]
You are right. I am sorry. I will fix that.

**Comment by fabian-hiller** [comment]
v0.24.1 is available. Can you check if it works now?

**Comment by NotWorkingCode** [comment]
I'll leave my code so you don't have to rewrite it.

https://pastebin.com/4AvtfBBy

**Comment by fabian-hiller** [comment]
Strange. It works as it should for me:

![screenshot-github-issue](https://github.com/fabian-hiller/valibot/assets/35291865/37c78eb6-6d8f-4bb4-aa9f-3447a45706c9)

**Comment by NotWorkingCode** [comment]
Do you think it could be the version of node being used?
I use 18

**Comment by fabian-hiller** [comment]
I don't think so. I also use Node.js v18. It must be related to TypeScript (I use v5.2.2).

**Comment by NotWorkingCode** [comment]
I'm using 5.3.2

**Comment by fabian-hiller** [comment]
TypeScript v5.3.2 produces the same results in my editor. Can you clone the Valibot repo and copy and past the following into `library/playground.ts` and check the output type?

```ts
import * as v from './mod.ts';

export const ConfigSchema = v.object({
  logger: v.object({
    level: v.string(),
    format: v.optional(v.string(), '{icon} > {time} | {level} | {message}'),
    path: v.optional(v.nullable(v.string()), null),
  }),
});

export type Config = v.Output<typeof ConfigSchema>;
```

**Comment by NotWorkingCode** [comment]
On playground, I get the correct result, the same as yours
![image](https://github.com/fabian-hiller/valibot/assets/52665159/993b929f-1364-4571-b81f-ad091b983e27)

**Comment by NotWorkingCode** [comment]
I tried to build the project and use the assembled library in playground - the problem returned

![image](https://github.com/fabian-hiller/valibot/assets/52665159/d586b036-0a83-4f8d-8f80-d5d043668688)

Apparently, the problem is only observed in the assembled library

**Comment by fabian-hiller** [comment]
You can run `cd library && pnpm build` to build Valibot. After that you can use the assembled library by `import * as v from './dist/index.js';` in `playground.ts`. There are no bugs for me here either.

**Comment by NotWorkingCode** [comment]
> You can run `cd library && pnpm build` to build Valibot. After that you can use the assembled library by `import * as v from './dist/index.js';` in `playground.ts`. There are no bugs for me here either.

I did exactly that. After building with pnpm, I imported from dist and the error returned. I showed it in my second screenshot.

**Comment by fabian-hiller** [comment]
This is strange. I don't know what to do. I reinstalled Valibot in another project and tested the code there. Everything works as it should. Since I can't reproduce it, I have no way to debug it. 😐

**Comment by NotWorkingCode** [comment]
I created the project on a completely different machine and got the same result

I use tsconfig by default

```
{
  "compilerOptions": {
    "target": "es2016",
    "module": "commonjs",
    "esModuleInterop": true,
    "forceConsistentCasingInFileNames": true,
    "strict": true,
    "skipLibCheck": true
  }
}
```

My clear project:
[SquadTS.zip](https://github.com/fabian-hiller/valibot/files/13670005/SquadTS.zip)

![image](https://github.com/fabian-hiller/valibot/assets/52665159/2159f561-820e-4b00-a4e0-cdf79e870393)

**Comment by Sec-ant** [comment]
I can partially reproduce this problem in a WebStorm IDE.

![image](https://github.com/fabian-hiller/valibot/assets/10386119/35388a31-d1b2-41f4-bda9-184096b19edf)

(You can see the type of `path` is `string` not `string | undefined` in my repro, the correct type should be `string | null` though)

But I do think this is an IDE bug (or mis-configuration, I don't use WebStorm so I don't know). Moreover, if the `format` type is what it is reported, then `expectString(config.logger.format)` should also report errors but it does not, which further suggests the IDE type hinting is somewhat problematic:

![image](https://github.com/fabian-hiller/valibot/assets/10386119/cb5107b1-beed-4f17-8c04-c94bce4227b7)

If this is an IDE bug, this shouldn't affect the build process, and I can confirm no type check errors are emitted when I use `npx tsc`:

![image](https://github.com/fabian-hiller/valibot/assets/10386119/18273b46-250b-41cb-b06c-95234de549d5)

So you can still build your project without issues.

And everything works correctly under VSCode.

Here is my test code:

```ts
import * as v from "valibot";

export const ConfigSchema = v.object({
  logger: v.object({
    level: v.string(),
    format: v.optional(v.string(), "{icon} > {time} | {level} | {message}"),
    path: v.optional(v.nullable(v.string()), null),
  }),
});

declare function expectString(_: string): void;

declare const config: v.Output<typeof ConfigSchema>;

expectString(config.logger.format);
```

**Comment by NotWorkingCode** [comment]
@Sec-ant You're damn right. I switched to VS Code myself a few hours ago and noticed that the problem was solved.
This seems to be a problem with the WebStorm static analyzer.

**Comment by NotWorkingCode** [comment]
@fabian-hiller The problem was indeed solved in version 2.24.1

I'm sorry for wasting your time.
Thank you so much for the quick fix.

**Comment by fabian-hiller** [comment]
No problem. Thanks a lot that you found this bug!
